okhttp/src/main/java/com/squareup/okhttp/HttpUrl.java

```diff
@@ -0,0 +1,294 @@
+/*
+ * Copyright (C) 2015 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.okhttp;
+
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.URI;
+import java.net.URL;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * A <a href="https://url.spec.whatwg.org/">URL</a> with an {@code http} or {@code https} scheme.
+ *
+ * TODO: discussion on canonicalization
+ *
+ * TODO: discussion on encoding-by-parts
+ *
+ * TODO: discussion on this vs. java.net.URL vs. java.net.URI
+ */
+public final class HttpUrl {
+  private HttpUrl(Builder builder) {
+    throw new UnsupportedOperationException();
+  }
+
+  public URL url() {
+    throw new UnsupportedOperationException();
+  }
+
+  public URI uri() throws IOException {
+    throw new UnsupportedOperationException();
+  }
+
+  /** Returns either "http" or "https". */
+  public String scheme() {
+    throw new UnsupportedOperationException();
+  }
+
+  public boolean isHttps() {
+    return scheme().equals("https");
+  }
+
+  public String user() {
+    throw new UnsupportedOperationException();
+  }
+
+  public String encodedUser() {
+    throw new UnsupportedOperationException();
+  }
+
+  /**
+   * Returns the decoded (potentially non-ASCII) hostname. The returned string may contain non-ASCII
+   * characters and is <strong>not suitable</strong> for DNS lookups; for that use {@link
+   * #encodedHost}. For example, this may return {@code ☃.net} which is a user-displayable IDN that
```

Comment:
It's not full-unicode. There's a full spec on case folding to avoid character sequences that are visually similar. Will be hard but fun to implement. http://www.unicode.org/reports/tr46/#Introduction